### ISSUE TITLE
Form edition do not handle all inputs correctly

### ISSUE THREAD

**Issue by eric-burel** [VulcanJS/Vulcan#1884]
Hi,

When editing a form, I experience some issues with custom controls. 

When I submit the form after editing some fields, the fields I did not edit (not even touch, so `onBlur` is not called) are now considered `$unset`.

This is related to the `onSubmit` callback of the Form component [there](https://github.com/VulcanJS/Vulcan/blob/master/packages/vulcan-forms/lib/components/Form.jsx#L579-L588).

The  `data` object configured just before is incorrect.So when you compute the fields, some of them are missing, and they end up in the `unset` field even if I did not delete the value.

I can't tell what function is responsible, there are a few calls to define this `data` object. My guess rn is that [when you define the `data` object](https://github.com/VulcanJS/Vulcan/blob/master/packages/vulcan-forms/lib/components/Form.jsx#L555) the first time, you rely on Formsy, which actually misses some controls.

Also a side note: the `FormComponent` inputs are uncontrolled (rely on `onBlur` instead of `onChange`) but when editing they got a `value`. React should have warned, but since there is a callback `onChange` that counts chars, it thinks it's ok. We should either should uncontrolled or controlled.

According to @ErikDakoda  and from React docs controlled should be preferred, however I don't know which is the longest to implement. Controlled can also have perf issues if you do heavy computations, like validation, on each keypress.

**Comment by eric-burel** [comment]
Components like the form upload use `addToAutofilledValues`  during mount:   `this.context.addToAutofilledValues({[this.props.name]: newValue});`. That is why they work as expected.

This call is actually mandatory for custom controls to work as expected during edit. So there seems to be 2 solutions:
- complete the doc about custom controls
- make this call unnecessary

If the user never changed the value of a given control through `updateCurrentValues`, it should not be considered as `unset`, but instead be `set` at its current value or ignored. I don't know how to implement this, maybe `this.autofilledValues` should be initialized differently in edit mode ?

**Comment by vincro** [comment]
One thing I do on my forms is to pre-fill the values by passing the document to SS to generate a default set of values, and then strap this to the form.  I have a custom setup, so not sure how this might work with SmartForm instead, but seems like you are most of the way there already.

**Comment by eric-burel** [comment]
Sadly it does not seem to work for objects :/

My `autofilledValues` is correctly set to `{ address: {...}}` in the `Form` component, but then the `address` field is still unset during submit. I am trying to figure out why.
It works for other custom components though.

**Comment by SachaG** [comment]
I have to confess I haven't had time to take another look at the form code recently. If this can be fixed easily I'd accept a PR, but if it requires a bigger refactoring then it'll probably have to wait a little…

**Comment by eric-burel** [comment]
Yes of course, it would even be a better idea to wait a little, because those issues might disappear by themselves next time someone works back on the forms. I'd like to help but I need a better understanding of how Formsy works first, I think this might need some non trivial refactor. In the meantime it is not that hard to quickfix using `updateCurrentValues`.
I'll try to update the `example-forms`.

**Comment by vincro** [comment]
I am willing to throw some time at this, just need a good "broken" example to review.  @eric-burel any chance of a repo to pull down and test against?

**Comment by eric-burel** [comment]
Thanks @vincro , that would be nice. Actually the bug should be reproducible in the `example-forms` I created in the Vulcan-starter (in a PR not merged yet), it concerns the `ObjectForm`. 

If you don't touch an address field during an edit, you will observe this issue (you may need to add a DataTable to see the data, I did not add one in the example).

**Comment by vincro** [comment]
@eric-burel will grab the merging branch and review later on today.

**Comment by SachaG** [comment]
> Also a side note: the FormComponent inputs are uncontrolled (rely on onBlur instead of onChange) but when editing they got a value. React should have warned, but since there is a callback onChange that counts chars, it thinks it's ok. We should either should uncontrolled or controlled.

I was revisiting this today and now I remember that the reason I used `onBlur` instead of `onChange` is because the Formsy inputs became super slow when I tried `onChange`. I agree `onBlur` is not great though because there are edge cases where you might submit a form without triggering the blur event. I'm not sure what the best solution is here…

**Comment by SachaG** [comment]
For now I think I'll use onChange, but debounce it. For some reason I can't get Formsy's debounce to work so I'll just do it at Vulcan's level.

**Comment by eric-burel** [comment]
Hi @SachaG, the lack of performance might also come from unexpected renders fired by rapid changes due to `onChange`, so you may want to check if debouncing actually solve the slowness issue on a non-trivial example.
I'll work back on forms when I have more time, I'll have a clearer agenda/roadmap in the weeks to come.
